server/src/test/java/io/crate/types/DataTypesTest.java

```diff
@@ -242,4 +242,111 @@ private static void assertCompareValueTo(DataType dt, Object val1, Object val2,
     public void test_can_guess_float_vectors() throws Exception {
         assertThat(DataTypes.guessType(new float[] { 3.14f })).isEqualTo(FloatVectorType.INSTANCE_ONE);
     }
+
+    @Test
+    public void test_merge_method_with_primitive_types() {
+        assertThat(DataTypes.merge(DataTypes.INTEGER, DataTypes.STRING)).isEqualTo(DataTypes.INTEGER);
+        assertThatThrownBy(() -> DataTypes.merge(DataTypes.IP, DataTypes.DATE))
+            .isExactlyInstanceOf(IllegalArgumentException.class)
+            .hasMessage("'date' is not convertible to 'ip'");
+    }
+
+    @Test
+    public void test_merge_method_with_primitive_type_and_container_type() {
+        assertThatThrownBy(() -> DataTypes.merge(DataTypes.INTEGER_ARRAY, DataTypes.INTEGER))
+            .isExactlyInstanceOf(IllegalArgumentException.class)
+            .hasMessage("'integer' is not convertible to 'integer_array'");
+        assertThatThrownBy(() -> DataTypes.merge(DataTypes.UNTYPED_OBJECT, DataTypes.INTEGER))
+            .isExactlyInstanceOf(IllegalArgumentException.class)
+            .hasMessage("'integer' is not convertible to 'object'");
+    }
+
+    @Test
+    public void test_merge_method_with_object_types() {
+        assertThat(
+            DataTypes.merge(
+                ObjectType.builder()
+                    .setInnerType("x", DataTypes.INTEGER)
+                    .setInnerType("y", DataTypes.STRING)
+                    .setInnerType("z", DataTypes.CHARACTER)
+                    .build(),
+                ObjectType.builder()
+                    .setInnerType("w", DataTypes.SHORT)
+                    .setInnerType("x", DataTypes.STRING)
+                    .setInnerType("y", DataTypes.INTEGER)
+                    .build()))
+            .isEqualTo(
+                ObjectType.builder()
+                    .setInnerType("w", DataTypes.SHORT)
```

Comment:
just double checking: is the order of the sub-cols deterministic (avoid flakiness)?